README
= pyexist =

pyexist is a simple Python wrapper around the eXist-db REST API.

For documentation please refer to

  [http://wiki.github.com/knipknap/pyexist/]

= Installation =

    sudo python setup.py install --prefix /usr/local

= Dependencies =

* Python 2.5 or greater
* Exist DB [http://www.exist-db.org/]


debian/changelog
pyexist (0.0) unstable; urgency=low

  * Initial release.

 -- Samuel Abels <knipknap@gmail.com>  Fri, 22 Jan 2010 22:15:08 +0100


version.sh
#!/bin/sh
# Tag revisions like this:
# $ git tag -a -m "v0.2" v0.2
VERSION_IN=VERSION.in
VERSION_FILE=src/pyexist/version.py

# Check that we are actually in a git managed project.
if [ ! -e .git -a -z "$1" ]; then
  echo >&2 Not a git repository.
  exit 1
fi

# Make sure that we have permission to modify the version file.
if [ -r $VERSION_FILE -a ! -w $VERSION_FILE ]; then
  echo >&2 No permission to modify $VERSION_FILE.
  exit 1
fi

# By default, get the version number from "git describe".
if [ ! -z "$1" ]; then
  VERSION=$1
else
  HEAD=`git log -1 --pretty=format:%H HEAD`
  VERSION=`git describe $HEAD --tags --match "v[0-9]*" | sed 's/-/./' 2>/dev/null`
  if [ -z "$VERSION" ]; then
    echo >&2 No matching tag was found.
    exit 1
  fi
fi

# If the --reset switch was given, reset the version number to 'DEVELOPMENT'.
[ "$1" = "--reset" ] && VERSION='DEVELOPMENT'

# If there is no version file, we are already done.
echo Version is $VERSION
[ ! -r $VERSION_FILE ] && exit 0

# Check whether the version file already contains this number,
# and only touch it if there is a change to avoid changing
# the timestamp.
VERSION_FILE_TMP=`tempfile`
cat $VERSION_IN | sed "s/@VERSION@/$VERSION/g" > $VERSION_FILE_TMP
if diff -q $VERSION_FILE_TMP $VERSION_FILE; then
  echo Version file unchanged.
  exit 0
fi

mv $VERSION_FILE_TMP $VERSION_FILE
echo Version file updated.


src/pyexist/XQuery.py
# Copyright (C) 2010 Samuel Abels.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2, as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
from lxml import etree

def _escape(arg):
    return str(arg).replace(r"'", r"''")

def _replacetags(string, **kwargs):
    for key, value in kwargs.iteritems():
        string = string.replace('%{' + key + '}', _escape(value))
    return string

class XQuery(object):
    """
    Represents a query. You normally don't want to create an XQuery instance
    directly, try using ExistDB.query() instead.
    
    Query evaluation is lazy, so the query is not executed before the result
    is requested using either the slice notation (such as query[0], or
    query[1:20]) or calling one of the count() or length methods.
    """
    def __init__(self, db, query, **kwargs):
        """
        Use ExistDB.query() instead of creating a query directly.

        @type  db: ExistDB
        @param db: The parent database instance.
        @type  query: string
        @param query: The xquery as a string.
        @type  kwargs: dict
        @param kwargs: Parameters to pass into the query.
        """
        self.db    = db
        self.query = _replacetags(query, **kwargs)
        self.len   = None

    @staticmethod
    def fromfile(db, filename, **kwargs):
        """
        Like the constructor, but reads the query from the file with the given
        name instead.

        @type  db: ExistDB
        @param db: The parent database instance.
        @type  filename: string
        @param filename: The name of a file containing an xquery statement.
        @type  kwargs: dict
        @param kwargs: Parameters to pass into the query.
        """
        return XQuery(db, open(file).read(), **kwargs)

    def __iter__(self):
        """
        Iterate over all results.

        @rtype:  iterator
        @return: An iterable the walks over all results.
        """
        return self[:].__iter__()

    def __len__(self):
        """
        Returns the number of matches that the query produces.

        @rtype:  long
        @return: The number of rows returned by the query.
        """
        return self.count()

    def count(self):
        """
        Equivalent to len().

        @rtype:  long
        @return: The number of rows returned by the query.
        """
        if self.len is None:
            # Execute the query with minimal results to see the count.
            tree     = self[0:1]
            self.len = int(tree.get('{' + self.db.RESULT_NS + '}hits'))
        return self.len

    def __getitem__(self, key):
        """
        Returns the range of matching items.

        @rtype:  lxml.etree._Element
        @return: The XML tree that is produced by the query.
        """
        # Parse the slice argument.
        if isinstance(key, int):
            start = int(key) + 1
            max   = 1
        elif isinstance(key, slice):
            # Try not to use key.indices(self.count()), as that would require
            # an extra query for counting the items.
            start = key.start is not None and key.start or 0
            step  = key.step or 1
            if step != 1:
                raise TypeError('slice step %d is not supported' % step)
            if key.stop is None:
                max = None
            else:
                max = key.stop - start
            start = start + 1
        else:
            raise TypeError('invalid key argument ' + repr(key))

        # Execute the query and parse the response.
        result = self.db._post(self.query, start = start, max = max)
        tree   = etree.fromstring(result)

        # Catch errors.
        if tree.tag == 'exception':
            try:
                error = tree.find('message').text
            except AttributeError:
                error = etree.tounicode(tree)
            raise self.db.Error('server said: ' + error)

        self.len = int(tree.get('{' + self.db.RESULT_NS + '}hits'))
        return tree


src/pyexist/ExistDB.py
# Copyright (C) 2010 Samuel Abels.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2, as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
import os, httplib, urlparse, base64
from XQuery import XQuery

_query_tmpl = '''
<query xmlns="http://exist.sourceforge.net/NS/exist"%s>
  <text><![CDATA[ %s ]]></text>
  <properties>
    <property name="indent" value="yes"/>
    <property name="pretty-print" value="yes"/>
  </properties>
</query>
'''

class ExistDB(object):
    """
    The eXist-db connection object.
    """
    RESULT_NS = 'http://exist.sourceforge.net/NS/exist'

    class Error(Exception):
        pass

    def __init__(self, host_uri, collection):
        """
        Create a new database connection using the REST protocol.

        @type  host_uri: string
        @param host_uri: The host and port number, separated by a ':' character.
        @type  collection: string
        @param collection: A database (collection) name.
        """
        # Python's urlparse module is so bad it hurts.
        uri = urlparse.urlparse('http://' + host_uri)
        try:
            auth, netloc = uri.netloc.split('@', 1)
        except ValueError:
            auth   = ''
            netloc = uri.netloc
        self.username   = auth.split(':', 1)[0]
        self.password   = auth[len(self.username) + 1:]
        self.collection = collection
        self.conn       = httplib.HTTP(netloc)
        self.path       = collection

    def _authenticate(self):
        if not self.username:
            return
        if self.password:
            auth = self.username + ':' + self.password
        else:
            auth = self.username
        auth = base64.encodestring(auth).strip()
        self.conn.putheader('Authorization', 'Basic ' + auth)

    def store(self, doc, xml):
        """
        Imports the XML string into the document with the given name.

        @type  doc: string
        @param doc: A document name.
        @type  xml: string
        @param xml: The XML to import.
        """
        self.conn.putrequest('PUT', self.path + '/' + doc)
        self._authenticate()
        self.conn.putheader('Content-Type',   'text/xml')
        self.conn.putheader('Content-Length', str(len(xml)))
        self.conn.endheaders()
        self.conn.send(xml)

        errcode, errmsg, headers = self.conn.getreply()
        if errcode != 201:
            raise ExistDB.Error('Error %d: %s' % (errcode, errmsg))

    def store_file(self, filename, doc = None):
        """
        Like store(), but reads the XML from a file instead. If the document
        name is None, it defaults to the basename of the file, with the .xml
        extension removed.

        @type  filename: string
        @param filename: The name of an XML file.
        @type  doc: string
        @param doc: A document name.
        """
        if doc is None:
            doc = os.path.splitext(os.path.basename(filename))[0]
        xml = open(filename).read()
        self.store(doc, xml)

    def delete(self, doc):
        """
        Deletes the document with the given name. Raises an error if the
        document does not exist.

        @type  doc: string
        @param doc: A document name.
        """
        self.conn.putrequest('DELETE', self.path + '/' + doc)
        self._authenticate()
        self.conn.endheaders()

        errcode, errmsg, headers = self.conn.getreply()
        if errcode != 200:
            raise ExistDB.Error('Error %d: %s' % (errcode, errmsg))

    def _post(self, thequery, start = 1, max = None):
        args = ''
        if start is not None:
            args += ' start="%d"' % start
        if max is None:
            args += ' max="-1"'
        else:
            args += ' max="%d"' % max
        thequery = _query_tmpl % (args, thequery)
        self.conn.putrequest('POST', self.path)
        self._authenticate()
        self.conn.putheader('Content-Type',   'text/xml')
        self.conn.putheader('Content-Length', str(len(thequery)))
        self.conn.endheaders()
        self.conn.send(thequery)

        errcode, errmsg, headers = self.conn.getreply()
        if errcode not in (200, 202):
            raise ExistDB.Error('Error %d: %s' % (errcode, errmsg))

        response = self.conn.getfile().read()
        return response

    def query(self, thequery, **kwargs):
        """
        Creates a new query object from the given xquery statement.
        The given kwargs are parameters that are replaced in the query.
        The query may use the following syntax for such parameters::

            let myvar := '%{myparam}'

        Passing "myparam = 'foo'" will produce the following query::

            let myvar := 'foo'

        @type  thequery: string
        @param thequery: The xquery as a string.
        @type  kwargs: dict
        @param kwargs: Parameters to pass into the query.
        @rtype:  XQuery
        @return: An XQuery object.
        """
        return XQuery(self, thequery, **kwargs)

    def query_from_file(self, filename, **kwargs):
        """
        Like query(), but reads the xquery from the file with the given
        name instead.

        @type  filename: string
        @param filename: The name of a file containing the query.
        @type  kwargs: dict
        @param kwargs: Parameters to pass into the query.
        @rtype:  XQuery
        @return: An XQuery object.
        """
        thequery = open(filename, 'r').read()
        return self.query(thequery, **kwargs)
